### PR TITLE
SendInBlue Nodes

### PR DESCRIPTION
**Pull request by brianinoa**
- SendInBlue Declarative Node
   - https://linear.app/n8n/issue/N8N-3789/sendinblue-node-implementation
- SendInBlue Trigger Node
   - https://linear.app/n8n/issue/N8N-3856/sendinblue-trigger-node

### CODE REVIEW

File: packages/nodes-base/nodes/SendInBlue/GenericFunctions.ts [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+		url: string,
+	): Promise<WebhookId> => {
+		const { apiKey } = await ref.getCredentials(credentialsName);
+		const endpoint = `${baseURL}/webhooks`;
+
+		const options: OptionsWithUri = {
+			method: 'POST',
+			headers: {
+				Accept: 'application/json',
+				'Api-Key': apiKey,
```
Comment (janober) [inline packages/nodes-base/nodes/SendInBlue/GenericFunctions.ts:189]: Why do we add here and in other locations the API key to the header? That kind of defeats the purpose of using "requestWithAuthentication" which exists to do exactly that and so ensures that authentication behaves consinstently.

File: packages/nodes-base/nodes/SendInBlue/SendInBlue.node.ts [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+import { INodeType, INodeTypeDescription } from 'n8n-workflow';
+import { attributeFields, attributeOperations } from './AttributeDescription';
+import { contactFields, contactOperations } from './ContactDescription';
+import { emailFields, emailOperations } from './EmailDescription';
+import { senderFields, senderOperations } from './SenderDescrition';
+
+export class SendInBlue implements INodeType {
+	description: INodeTypeDescription = {
+		displayName: 'SendInBlue',
+		name: 'sendinblue',
```
Comment (janober) [inline packages/nodes-base/nodes/SendInBlue/SendInBlue.node.ts:11]: The name is wrong. It has to be "sendInBlue".

File: packages/nodes-base/credentials/SendInBlueApi.credentials.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
+import {
+	IAuthenticateGeneric,
+	ICredentialTestRequest,
+	ICredentialType,
+	INodeProperties,
+} from 'n8n-workflow';
+
+export class SendInBlueApi implements ICredentialType {
+	name = 'sendinblueApi';
```
Comment (janober) [inline packages/nodes-base/credentials/SendInBlueApi.credentials.ts:9]: The name is wrong. Would have to be "sendInBlueApi". Maybe also check with @RicardoE105 I thought we do not use the "Api" at the end anymore.

File: packages/nodes-base/credentials/SendInBlueApi.credentials.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+	displayName = 'SendInBlue API';
+	documentationUrl = 'sendinblue';
```
Comment (janober) [inline packages/nodes-base/credentials/SendInBlueApi.credentials.ts:11]: Same as above

File: packages/cli/src/Logger.ts [hunk 5]
```diff
@@ -4,6 +4,7 @@
 /* eslint-disable @typescript-eslint/no-unsafe-return */
 
 import { inspect } from 'util';
+
```
Comment (janober) [inline packages/cli/src/Logger.ts:7]: Guess we can remove that. No need to have it as part of this PR.

File: packages/nodes-base/nodes/SendInBlue/SendInBlueTrigger.node.ts [hunk 6]
```diff
@@ -0,0 +1,300 @@
...
+		displayName: 'SendInBlue Trigger',
+		defaults: {
+			name: 'SendInBlue-Trigger',
+			color: '#044a75',
+		},
+		description: 'Starts the workflow when SendInBlue events occur',
+		group: ['trigger'],
+		icon: 'file:sendinblue.svg',
+		inputs: [],
+		name: 'sendinblueTrigger',
```
Comment (janober) [inline packages/nodes-base/nodes/SendInBlue/SendInBlueTrigger.node.ts:30]: Name is wrong. Should be "sendInBlueTrigger"

File: packages/nodes-base/nodes/SendInBlue/SendInBlueTrigger.node.ts [hunk 7]
```diff
@@ -0,0 +1,300 @@
...
+		outputs: ['main'],
+		version: 1,
+		webhooks: [
+			{
+				name: 'default',
+				httpMethod: 'POST',
+				responseMode: 'onReceived',
+				path: 'webhooks',
+				responseData: 'allEntries',
```
Comment (janober) [inline packages/nodes-base/nodes/SendInBlue/SendInBlueTrigger.node.ts:39]: This option can probably be removed. As it is returns the data "onReceived" and there are so no entries to return.

File: packages/nodes-base/nodes/SendInBlue/SendInBlueTrigger.node.ts [hunk 8]
```diff
@@ -0,0 +1,300 @@
...
+		properties: [
+			{
+				default: 'transactional',
+				displayName: 'Resource',
+				name: 'type',
+				options: [
+					{ name: 'Transactional', value: 'transactional' },
+					{ name: 'Marketing', value: 'marketing' },
+					{ name: 'Inbound', value: 'inbound' },
+				],
```
Comment (janober) [inline packages/nodes-base/nodes/SendInBlue/SendInBlueTrigger.node.ts:51]: Should be sorted alphabetically.

File: packages/nodes-base/nodes/SendInBlue/SendInBlue.node.ts [hunk 9]
```diff
@@ -0,0 +1,74 @@
...
+					},
+					{
+						name: 'Sender',
+						value: 'sender',
+					},
+					{
+						name: 'Contact Attribute',
+						value: 'attribute',
+					},
+				],
```
Comment (janober) [inline packages/nodes-base/nodes/SendInBlue/SendInBlue.node.ts:60]: Should be sorted alphabetically.

File: packages/nodes-base/nodes/SendInBlue/SendInBlue.node.ts [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+				name: 'sendinblueApi',
+				required: true,
+			},
+		],
+		requestDefaults: {
+			baseURL: 'https://api.sendinblue.com',
+			url: '',
+			headers: {
+				Accept: 'application/json',
+				'Content-Type': 'application/json',
```
Comment (janober) [inline packages/nodes-base/nodes/SendInBlue/SendInBlue.node.ts:34]: Is this really needed? Do we not do that by default?

File: packages/nodes-base/credentials/SendInBlueApi.credentials.ts (hunk 3 above)
Comment (brianinoa) [reply packages/nodes-base/credentials/SendInBlueApi.credentials.ts:9]: @janober 
It looks to be in the auto-fixable [lint rules](https://github.com/ivov/eslint-plugin-n8n-nodes-base/blob/master/docs/rules/cred-class-field-name-unsuffixed.md)

**Review by brianinoa — COMMENTED** [review]
@janober thanks for the feedback.

All comments have been addressed, as well as other issues I uncovered

File: packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts [hunk 11]
```diff
@@ -0,0 +1,531 @@
...
+			},
+			{
+				name: 'Float',
+				value: 'float',
+			},
+			{
+				name: 'Boolean',
+				value: 'boolean',
+			},
+		],
```
[the remark this answers is not in the document]
Comment (ivov) [reply packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts]: @janober [Min number for alphabetization is 5](https://github.com/ivov/eslint-plugin-n8n-nodes-base/blob/4f8bf254845bb0045c45edc3dbebc0c29e6fb1ab/lib/constants.ts#L9), set by David. Should it be lower?

File: packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts [hunk 12]
```diff
@@ -0,0 +1,531 @@
...
+				resource: ['attribute'],
+				operation: ['createAttribute'],
+			},
+		},
+		required: true,
+		name: 'attributeName',
+		type: 'string',
+	},
+	{
+		default: '',
```
[the remark this answers is not in the document]
Comment (ivov) [reply packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts:186]: https://github.com/ivov/eslint-plugin-n8n-nodes-base/issues/93

File: packages/nodes-base/credentials/SendInBlueApi.credentials.ts (hunk 3 above)
Comment (ivov) [reply packages/nodes-base/credentials/SendInBlueApi.credentials.ts]: https://github.com/ivov/eslint-plugin-n8n-nodes-base/issues/94

File: packages/nodes-base/nodes/SendInBlue/SendInBlueTrigger.node.ts (hunk 8 above)
Comment (ivov) [reply packages/nodes-base/nodes/SendInBlue/SendInBlueTrigger.node.ts]: Ref: https://github.com/n8n-io/n8n/pull/3746#discussion_r930197999

File: packages/nodes-base/nodes/SendInBlue/SendInBlue.node.ts (hunk 9 above)
Comment (ivov) [reply packages/nodes-base/nodes/SendInBlue/SendInBlue.node.ts]: Ref: https://github.com/n8n-io/n8n/pull/3746#discussion_r930195753

File: packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts [hunk 13]
```diff
@@ -0,0 +1,531 @@
...
+
+								return items;
+							},
+						],
+					},
+				},
+				action: 'Get all attributes',
+			},
+		],
+		default: 'createAttribute',
```
Comment (krynble) [inline packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts:129]: Please also change the default here to `create`

File: packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts [hunk 14]
```diff
@@ -0,0 +1,531 @@
...
+const updateAttributeOperations: INodeProperties[] = [
+	{
+		displayName: 'Category',
+		name: 'updateAttributeCategory',
+		default: 'calculated',
+		description: 'Category of the attribute',
+		displayOptions: {
+			show: {
+				resource: ['attribute'],
+				operation: ['updateAttribute'],
```
Comment (krynble) [inline packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts:319]: Change to `update` - needs to also change the other Update related fields.

File: packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts [hunk 15]
```diff
@@ -0,0 +1,531 @@
...
+				default: {},
+				description: 'List of values and labels that the attribute can take',
+			},
+		],
+	},
+];
+
+const deleteAttribueOperations: INodeProperties[] = [
+	{
+		displayName: 'Category',
```
Comment (krynble) [inline packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts:443]: Is the category necessary when removing an attribute?

File: packages/nodes-base/nodes/SendInBlue/SenderDescrition.ts [hunk 16]
```diff
@@ -0,0 +1,196 @@
...
+								type: 'rootProperty',
+								properties: {
+									property: 'senders',
+								},
+							},
+							async function (
+								this: IExecuteSingleFunctions,
+								items: INodeExecutionData[],
+							): Promise<INodeExecutionData[]> {
+								const returnAll = this.getNodeParameter('returnAll') as boolean;
```
Comment (krynble) [inline packages/nodes-base/nodes/SendInBlue/SenderDescrition.ts:64]: Isn't it possible to send this as a request param to fetch the exact number already? If not, you can disregard this comment.

**Review by krynble — CHANGES_REQUESTED** [review]
I've found some more issues, can you please fix them @brianinoa ?

Afterwards I think it's worth running a quick check on all resources and operations just to make sure it's 100%, I haven't done so in this run.

File: packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts (hunk 15 above)
Comment (brianinoa) [reply packages/nodes-base/nodes/SendInBlue/AttributeDescription.ts:443]: Yes it's required, I updated the field to be `{required:true}`

File: packages/nodes-base/nodes/SendInBlue/SenderDescrition.ts (hunk 16 above)
Comment (brianinoa) [reply packages/nodes-base/nodes/SendInBlue/SenderDescrition.ts:64]: No, they actually don't support query parameters on that api call

[getSenders](https://developers.sendinblue.com/reference/getsenders-1)

**Comment by janober** [comment]
Thanks a lot. Got merged.

**Comment by janober** [comment]
Got released with `n8n@0.189.0`

**Pull request merged**